### PR TITLE
feat(article): diagnostics channel landed!

### PR DESCRIPTION
**Pull request by RafaelGSS**
A quick introduction to `diagnostics_channel` and a try to call open-source developers to help with support.

### CODE REVIEW

File: priv/posts/08-diagnostics-channel-is-landed.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+For instance, we can have:
+
+- Tracking entire http lifecycle
+- Measure core metrisc of our application easier (event-loop, gc, memory utilization, cpu utilization)
+
+Of course, this feature sounds better for APM Vendors.
+
+## Why use this module instead of EventEmmiter?
+
+// TODO
```
Comment (RafaelGSS) [inline priv/posts/08-diagnostics-channel-is-landed.md:99]: @Qard, can you help here?

File: priv/posts/08-diagnostics-channel-is-landed.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+# Diagnostics Channel
+
+In `node` v15 was landed a feature that should help a lot of APM vendors. `diagnostics_channel` has the objective to have a centralized channel of events between modules.
+
+As blog release says:
+
+> _`diagnostics_channel` is a new experimental module that provides an API to create named channels to report arbitrary message data for diagnostics purposes.
+With `diagnostics_channel`, Node.js core and module authors can publish contextual data about what they are doing at a given time. This could be the hostname and query string of a mysql query, for example. Just create a named channel with dc.channel(name) and call channel.publish(data) to send the data to any listeners to that channel._
+
+This feature is similar to [`EventEmmiter`](https://nodejs.org/api/events.html#events_class_eventemitter), however is optimized enough to propagate data synchronous.
```
Comment (Qard) [inline priv/posts/08-diagnostics-channel-is-landed.md:16]: `EventEmitter` is sync too. A more accurate statement would be that `diagnostics_channel` does have the inherent publishing cost that emitting a string-named event to an `EventEmitter` has. With `EventEmitter` there is a lookup cost to every single publish, mapping the string name of the event to the listener set. With `diagnostics_channel`, the separate object scoping means that the publish points _directly_ to the memory of the handler list.

File: priv/posts/08-diagnostics-channel-is-landed.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+APM Vendors usually does `monkey-patch` of every key modules for publish information. In `diagnostics_channel` world, we can avoid it in favor of events.
```
Comment (Qard) [inline priv/posts/08-diagnostics-channel-is-landed.md:18]:
```suggestion
APM Vendors usually monkey patch every key module to capture the information which `diagnostics_channel` simply publishes as events. Monkey patching generally creates many additional costly closures and can be fragile so it's much safer to rely on more intentional events.
```

File: priv/posts/08-diagnostics-channel-is-landed.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+Actually, most of the database modules already emit the warning on query slow if you set on the settings. This example is just to show you the main usage of this module.
+
+After key modules support `diagnostics_channel` we are able to have a better observability/tracing of our application without add a lot of complexity in our nodejs code.
+
+For instance, we can have:
+
+- Tracking entire http lifecycle
+- Measure core metrisc of our application easier (event-loop, gc, memory utilization, cpu utilization)
```
Comment (Qard) [inline priv/posts/08-diagnostics-channel-is-landed.md:93]: I think you would generally capture that particular examples of metrics in other ways. You do something like count requests though.

File: priv/posts/08-diagnostics-channel-is-landed.md (hunk 1 above)
Comment (Qard) [reply priv/posts/08-diagnostics-channel-is-landed.md:99]: `EventEmitter` has an extra cost on every single publish to look up the handler set by the string event name. That's not a huge deal for just a single run, but in a high-frequency scenario where the logic might be repeated thousands, or even _millions_, of times per second, it adds up _fast_. Additionally, the lookup cost _always_ happens with `EventEmitter` while with `diagnostics_channel` it _only_ happens when something is actually _listening_ to that specific channel. The intent is for there to be _hundreds_ or even _thousands_ of these channels being reported to at any given time while there might be only a _few_ of those channels being actively _observed_ at any given time. The majority of the time there would be nothing to publish to so it's been intentionally designed to do _nothing at all_ in that case. This design makes it much more suitable as a _data firehose_ whereas a typical `EventEmitter` is really only suited to a more _limited_ set of events.

File: priv/posts/08-diagnostics-channel-is-landed.md (hunk 4 above)
Comment (RafaelGSS) [reply priv/posts/08-diagnostics-channel-is-landed.md:93]: Not sure, the APM vendors will publish this kind of data through dc channels as far I know.

Comment (Qard) [reply priv/posts/08-diagnostics-channel-is-landed.md:95]: Those particular types of metrics don't really have a trigger point that makes it make sense for diagnostics_channel. They would generally be collected through some polling mechanism, like have a `setInterval` run `process.cpuUsage()` and `process.memoryUsage()` every X number of seconds. Don't need a channel for that. 🤷

Comment (RafaelGSS) [reply priv/posts/08-diagnostics-channel-is-landed.md:95]: Yes, you right. I thought about post-processing data around this type.

**Pull request merged**